iis-wf/iis-wf-export-actionmanager/src/main/java/eu/dnetlib/iis/wf/export/actionmanager/module/CitationsActionBuilderModuleFactory.java

```diff
@@ -18,101 +9,203 @@
+            if (source != null) {
+                Float confidenceLevelThreshold = trustLevelConverter.convert(trustLevelThreshold);
+                return source.stream()
+                        .map(citationEntry -> citationEntryNormalizer.normalize(citationEntry,
+                                citationEntryMatchValidator.validate(citationEntry, confidenceLevelThreshold)
+                        ))
+                        .map(citationEntry -> blobCitationEntryBuilder.build(citationEntry))
+                        .collect(Collectors.toCollection(TreeSet::new));
+            }
+            return null;
         }
 
-        // ------------------------ PRIVATE --------------------------
-        
         /**
-         * Performs confidence level normalization. Removes empty lists. 
-         * Removes 50| prefix from publication identifier.
-         * 
-         * @param source list of citations to be normalized
-         * @return {@link BlobCitationEntry} objects having confidence level value normalized
+         * Allows trust level threshold to be converted to confidence level threshold.
          */
-        private SortedSet<BlobCitationEntry> normalize(List<CitationEntry> source) {
-            if (source != null) {
-                SortedSet<BlobCitationEntry> results = new TreeSet<BlobCitationEntry>();
-                for (CitationEntry currentEntry : source) {
-                    if (currentEntry.getExternalDestinationDocumentIds().isEmpty()) {
-                        currentEntry.setExternalDestinationDocumentIds(null);
-                    }
-                    if (currentEntry.getDestinationDocumentId() != null) {
-                        currentEntry.setDestinationDocumentId(
-                                StringUtils.split(currentEntry.getDestinationDocumentId().toString(),
-                                        InfoSpaceConstants.ROW_PREFIX_SEPARATOR)[1]);
-                    }
-                    results.add(CitationsActionBuilderModuleUtils.build(currentEntry));
+        public static class TrustLevelConverter {
+            private Float conversionFactor;
+
+            public TrustLevelConverter(Float conversionFactor) {
+                this.conversionFactor = conversionFactor;
+            }
+
+            public Float convert(Float trustLevelThreshold) {
+                return Optional.ofNullable(trustLevelThreshold)
+                        .map(x -> x / conversionFactor)
+                        .orElse(null);
+            }
+        }
+
+        /**
+         * Allows to validate citation entry match.
+         */
+        public static class CitationEntryMatchValidator {
+            private ConfidenceLevelValidator confidenceLevelValidator = new ConfidenceLevelValidator();
+
+            public Boolean validate(CitationEntry citationEntry, Float confidenceLevelThreshold) {
+                if (Objects.isNull(citationEntry.getConfidenceLevel())) {
+                    return false;
```

Comment:
This comment is partially applicable to the lines 109-112 but the condition logic I am referring to is part of the code I have marked here.

~~WARN: if I understand this properly this could rule out all the citation entries which were unmatched. We definitely DO NOT want this to happen!~~

I have studied the code more thoroughly and it seems we will not drop the unmatched entries from the list of bibrefs (it is not part of the filter in streaming processing). But still the logic here triggers unnecessary logic of nulling `destinationDocumentId` and `convidenceLevel` (which is already null but this is not the core of the problem) fields and suggests such entry is invalid (as the validator returns `false`) what may give impression there is something wrong with the record which is not the case: it simply does not convey the outcome of citation matching (this is why the confidence level is `null`) but it definitely brings the text and possible external ids we want to make a part of the BLOB.

We should focus on removing matches (or normalizing the matched ids) by applying the `CitationEntryNormalizer` but ONLY on matched entries (indicated by confidence level and destination document id presence). So we should refactor it in a way missing confidence level should not yield the negative outcome of validation process (but the positive outcome should not trigger `validMatchCitationEntryNormalizer`). We could rename the validator to be more adequate to what it is doing and apply norimalizer on matched entries only.

Reply:
I should've asked more questions while working on this task :/ 

I've refactored the code that was poorly designed in terms of checking if a citation entry is a result of citation matching. Firstly I've changed the name to 'checker' to show that we only perform a check. Secondly, I've added an enum containing the result of this check. Thirdly, I've refactored the processing flow to validate against threshold and to normalize only citation entries that contain a result of citation matching. 

I've compared the results of running the exporting workflow before and after above changes and I've got the same results, but the code now is much clearer.